File: Python/battlecityclient/internals/point.py
class Point(object):
    """ Describes a point on board."""
    def __init__(self, x=0, y=0):
        self._x = int(x)
        self._y = int(y)

    def __key(self):
        return self._x, self._y

    def __str__(self):
        return self.to_string()

    def __repr__(self):
        return self.to_string()

    def __eq__(self, other_point):
        return self.__key() == other_point.__key()

    def __hash__(self):
        return hash(self.__key())

    def get_x(self):
        return self._x

    def get_y(self):
        return self._y

    def abs_sum(self):
        return abs(self._x) + abs(self._y)

    def diff(self, o):
        return Point(x=self._x - o._x, y=self._y - o._y)

    def manhattan_distance(self, o):
        return self.diff(o).abs_sum()

    def to_string(self):
        return "[{},{}]".format(self._x, self._y)


if __name__ == '__main__':
    raise RuntimeError("This module is not expected to be ran from CLI")

File: java/src/main/java/clientlib/model/Point.java
package clientlib.model;


import clientlib.map.Type;
import lombok.*;

import java.io.Serializable;

import static clientlib.model.Direction.*;

@Data
@ToString
@EqualsAndHashCode
@NoArgsConstructor
@AllArgsConstructor
public class Point implements Serializable {

    protected int x;
    protected int y;
    protected Type type;

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }


    public Point diff(Point other) {
        return new Point(x - other.x, y - other.y);
    }

    public Point dir(Direction direction) {
        return new Point(x + direction.getDx(), y + direction.getDy());
    }

    public int absSum() {
        return Math.abs(x) + Math.abs(y);
    }

    public int manhattanDistance(Point other) {
        return diff(other).absSum();
    }

    public Direction toBiggestDir() {
        if (Math.abs(x) < Math.abs(y)) {
            return y > 0 ? UP : DOWN;
        }
        return x > 0 ? RIGHT : LEFT;
    }

    public boolean isSamePoint(Point other) {
        return x == other.getX() && y == other.getY();
    }

}
